### PR TITLE
Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::SortedArrayMap

### PR DESCRIPTION
**Pull request by cdumez**
#### 14845fd3196cc136e21837eb4e02c80c56953e1b
<pre>
Stop using WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::SortedArrayMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=283638">https://bugs.webkit.org/show_bug.cgi?id=283638</a>

Reviewed by Geoffrey Garen.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_objc_protocol_type_conversions_header.py:
(ObjCProtocolTypeConversionsHeaderGenerator._generate_enum_from_protocol_string):
* Source/WTF/wtf/SortedArrayMap.h:
(WTF::ComparableASCIISubsetLiteral&lt;subset&gt;::ComparableASCIISubsetLiteral):
(WTF::equalSpansConstExpr):
(WTF::lessThanASCIICaseFolding):
(WTF::operator==):
(WTF::operator&lt;):
(WTF::subset&gt;::PackedASCIISubsetLiteral):
(WTF::subset&gt;::pack):
(WTF::strcmpConstExpr): Deleted.
* Source/WTF/wtf/text/ASCIILiteral.h:
* Source/WebCore/Modules/WebGPU/GPUAdapter.cpp:
(WebCore::convertFeatureNameToEnum):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseDir):
(WebCore::ApplicationManifestParser::parseDisplay):
(WebCore::ApplicationManifestParser::parseOrientation):
* Source/WebCore/Modules/mediacapabilities/MediaCapabilities.cpp:
(WebCore::isValidMediaMIMEType):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateEnumerationImplementationContent):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
(WebCore::parseEnumerationFromString&lt;TestCallbackInterface::Enum&gt;):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONEnum.cpp:
(WebCore::parseEnumerationFromString&lt;TestDefaultToJSONEnum&gt;):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::parseEnumerationFromString&lt;TestObj::EnumType&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::EnumTrailingComma&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::Optional&gt;):
(WebCore::parseEnumerationFromString&lt;AlternateEnumName&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::EnumA&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::EnumB&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::EnumC&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::Kind&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::Size&gt;):
(WebCore::parseEnumerationFromString&lt;TestObj::Confidence&gt;):
* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneDictionary.cpp:
(WebCore::parseEnumerationFromString&lt;TestStandaloneDictionary::EnumInStandaloneDictionaryFile&gt;):
* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneEnumeration.cpp:
(WebCore::parseEnumerationFromString&lt;TestStandaloneEnumeration&gt;):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSCustomPropertySyntax.cpp:
(WebCore::CSSCustomPropertySyntax::typeForTypeName):
* Source/WebCore/html/Autofill.cpp:
* Source/WebCore/html/EnterKeyHint.cpp:
(WebCore::enterKeyHintForAttributeValue):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::tagIdFor):
* Source/WebCore/mathml/MathMLPresentationElement.cpp:
(WebCore::MathMLPresentationElement::parseMathVariantAttribute):
* Source/WebCore/page/PerformanceUserTiming.cpp:
* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
(WebCore::keyboardScrollingKeyForKeyboardEvent):
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::normalizedImageMIMEType):
* Source/WebCore/platform/audio/glib/MediaSessionGLib.cpp:
(WebCore::getCommand):
(WebCore::handleMethodCall):
(WebCore::getMprisProperty):
(WebCore::handleGetProperty):
(WebCore::handleSetProperty):
(WebCore::MediaSessionGLib::create):
(WebCore::MediaSessionGLib::ensureMprisSessionRegistered):
(WebCore::MediaSessionGLib::emitPositionChanged):
(WebCore::MediaSessionGLib::updateNowPlaying):
(WebCore::MediaSessionGLib::getMetadataAsGVariant):
(WebCore::MediaSessionGLib::emitPropertiesChanged):
(WebCore::MediaSessionGLib::playbackStatusChanged):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::hasValidAverageCharWidth const):
* Source/WebCore/platform/graphics/HEVCUtilities.cpp:
(WebCore::parseDoViCodecType):
(WebCore::profileIDForAlphabeticDoViProfile):
* Source/WebCore/platform/graphics/avfoundation/objc/AVAssetMIMETypeCache.mm:
(WebCore::AVAssetMIMETypeCache::isUnsupportedContainerType):
(WebCore::AVAssetMIMETypeCache::isStaticContainerType):
* Source/WebCore/platform/network/mac/UTIUtilities.mm:
(WebCore::UTIFromUnknownMIMEType):
* Source/WebCore/platform/network/mac/WebCoreURLResponse.mm:
(WebCore::adjustMIMETypeIfNecessary):
* Source/WebCore/platform/text/LocaleToScriptMapping.cpp:
(WebCore::scriptNameToCode):
(WebCore::localeToScriptCodeForFontSelection):
* Source/WebCore/svg/SVGComponentTransferFunctionElement.h:
(WebCore::SVGPropertyTraits&lt;ComponentTransferType&gt;::fromString):
* Source/WebCore/svg/SVGFECompositeElement.h:
(WebCore::SVGPropertyTraits&lt;CompositeOperationType&gt;::fromString):
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::toCommandHTTPMethod):
* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer&lt;T&gt;::nextToken):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::visitArguments):
(WGSL::Metal::emitTextureDimensions):
(WGSL::Metal::emitTextureGather):
(WGSL::Metal::emitTextureLoad):
(WGSL::Metal::emitTextureSampleGrad):
(WGSL::Metal::emitTextureSampleLevel):
(WGSL::Metal::emitTextureSampleBaseClampToEdge):
(WGSL::Metal::emitTextureSampleBias):
(WGSL::Metal::emitTextureStore):
(WGSL::Metal::atomicFunction):
(WGSL::Metal::emitPack4xI8Clamp):
(WGSL::Metal::emitPack4xU8Clamp):
(WGSL::Metal::FunctionDefinitionWriter::serializeBinaryExpression):
(WGSL::Metal::FunctionDefinitionWriter::serializeConstant):
* Source/WebGPU/WGSL/Types.h:
* Source/WebGPU/WGSL/WGSLEnums.cpp:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(allowListedClassToString):
(dumpUIView):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::commandNameForSelectorName):
* Tools/TestWebKitAPI/Tests/WTF/SortedArrayMap.cpp:
(TEST(WTF, SortedArraySet)):

Canonical link: <a href="https://commits.webkit.org/287076@main">https://commits.webkit.org/287076@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b5e66f46a3065470e4102297f310b13eab6ebfd5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78249 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31634 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82911 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29517 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80382 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66442 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5576 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61285 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19203 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81316 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51277 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/68324 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41602 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/48624 "Build was cancelled. Recent messages:OS: Sonoma (14.6), Xcode: 15.4") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24895 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27855 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71397 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69717 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84279 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77486 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5614 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3791 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69513 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5771 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67206 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68768 "103 api tests failed or timed out") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12767 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11039 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99796 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12096 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5563 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/8321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21793 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5552 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8984 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7341 "Built successfully") | | |

### CODE REVIEW

File: Source/WTF/wtf/SortedArrayMap.h [hunk 1]
```diff
@@ -230,51 +227,51 @@ template<typename ArrayType> template<typename KeyArgument> inline bool SortedAr
...
 
-template<typename CharacterType> inline bool lessThanASCIICaseFolding(std::span<const CharacterType> characters, const char* literalWithNoUppercase)
+template<typename CharacterType> inline bool lessThanASCIICaseFolding(std::span<const CharacterType> characters, ASCIILiteral literalWithNoUppercase)
 {
-    for (auto character : characters) {
-        auto literalCharacter = *literalWithNoUppercase;
-        if (!literalCharacter)
-            return false;
+    size_t index = 0;
+    for (auto character : characters.first(std::min(characters.size(), literalWithNoUppercase.length()))) {
```
Comment (geoffreygaren) [inline Source/WTF/wtf/SortedArrayMap.h]: I think you can use IndexedRange here, so you don't have to `++index` at the bottom of the loop.

File: Source/WTF/wtf/SortedArrayMap.h [hunk 2]
```diff
@@ -230,51 +227,51 @@ template<typename ArrayType> template<typename KeyArgument> inline bool SortedAr
...
 
-template<typename CharacterType> inline bool lessThanASCIICaseFolding(const char* literalWithNoUppercase, std::span<const CharacterType> characters)
+template<typename CharacterType> inline bool lessThanASCIICaseFolding(ASCIILiteral literalWithNoUppercase, std::span<const CharacterType> characters)
 {
-    for (auto character : characters) {
-        auto literalCharacter = *literalWithNoUppercase;
-        if (!literalCharacter)
-            return true;
+    size_t index = 0;
+    for (auto character : characters.first(std::min(characters.size(), literalWithNoUppercase.length()))) {
```
Comment (geoffreygaren) [inline Source/WTF/wtf/SortedArrayMap.h]: I think you can use IndexedRange here, so you don't have to `++index` at the bottom of the loop.

File: Source/WTF/wtf/SortedArrayMap.h [hunk 3]
```diff
@@ -230,51 +227,51 @@ template<typename ArrayType> template<typename KeyArgument> inline bool SortedAr
     return iterator != std::end(m_array) && *iterator == *parsedKey;
 }
 
-constexpr int strcmpConstExpr(const char* a, const char* b)
+constexpr int compareSpansConstExpr(std::span<const char> a, std::span<const char> b)
```
Comment (geoffreygaren) [inline Source/WTF/wtf/SortedArrayMap.h:231]: I would call this `strcmpSpans`. strcmp has the unique behavior that it truncates the comparison. `equalSpans`, on the other hand, compares not equal when they're not the same size.

**Review by geoffreygaren — APPROVED** [review]
r=me

File: Source/WTF/wtf/SortedArrayMap.h (hunk 3 above)
Comment (darinadler) [reply Source/WTF/wtf/SortedArrayMap.h:231]: I don't agree. The word "compare", whether shortened to "cmp" or not, means a 3 way comparison that can take length into account by sorting the lack of bytes before any specific byte values when one argument has a longer length than the other. I don't think the "str" is important to that semantic.

It's analogous to how "equal" can return false when two lengths are not equal.

**Comment by cdumez** [comment]
I'd like to perf A/B test before landing but A/B testing has been down. I'll delay landing this until I can perf test it,

File: Source/WTF/wtf/SortedArrayMap.h [hunk 4]
```diff
@@ -230,51 +228,47 @@ template<typename ArrayType> template<typename KeyArgument> inline bool SortedAr
...
-        ++a;
-        ++b;
-    }
-    return *a == *b ? 0 : *a < *b ? -1 : 1;
+    auto commonLength = std::min(a.size(), b.size());
+    size_t i = 0;
+    while (i < commonLength && a[i] == b[i])
+        ++i;
+    auto aCharacter = i < a.size() ? a[i] : 0;
+    auto bCharacter = i < b.size() ? b[i] : 0;
```
Comment (darinadler) [inline Source/WTF/wtf/SortedArrayMap.h]: This does not seem quite right. We don't want lack of a character to sort equal to a 0 character. The lack of a character should sort as less than any character. It was different when the strings were null-terminated.

Comment (darinadler) [inline Source/WTF/wtf/SortedArrayMap.h]:
```suggestion
    if (i == commonLength) {
        if (a.size() == b.size())
            return 0;
        return a.size() < b.size() ? -1 : 1;
    }
    auto aCharacter = a[i];
    auto bCharacter = b[i];
```

Comment (cdumez) [reply Source/WTF/wtf/SortedArrayMap.h]: The suggested code doesn't seem to match the original behavior?
```
if (a.size() != b.size())
        return a.size() < b.size() ? -1 : 1;
```

What was the point of finding the first character mismatch index in the above loop to then ignore it and then just compare the size. I think we're only supposed to compare the sizes if ALL the characters matched in the above loop.

Comment (cdumez) [reply Source/WTF/wtf/SortedArrayMap.h]: I was trying to match the original code. The original code was dropping out of the loop if we reached a character mismatch and or a null terminator.

Falling back to 0 here when `I` is larger then than span size was my attempt to keep the same behavior as with nun-terminated strings. The previous code was using 0 in the comparison if the loop above aborted due to reaching the null terminator.

Comment (darinadler) [reply Source/WTF/wtf/SortedArrayMap.h]: Yes I understand that you were preserving behavior. My code was wrong but what we will want here is something that treats a short string as less than a longer one with the same prefix with no special handling of null characters.

Comment (darinadler) [reply Source/WTF/wtf/SortedArrayMap.h]: Yes I coded it wrong, but we ideally don't want to treat the character past the end of the string as a 0 even though the old code did that. When the strings are null terminated we are guaranteed that no characters are 0. But when they are not we want to correctly handle null characters.

**Pull request merged**